qiskit/quantum_info/operators/scalar_op.py

```diff
@@ -0,0 +1,264 @@
+            BaseOperator: if other is not an ScalarOp.
+        """
+        if not isinstance(other, BaseOperator):
+            other = Operator(other)
+        if isinstance(other, ScalarOp):
+            if self.coeff is None:
+                coeff = other.coeff
+            elif other.coeff is None:
+                coeff = self.coeff
+            else:
+                coeff = self.coeff * other.coeff
+            dims = self._input_dims + other._input_dims
+            return ScalarOp(dims, coeff=coeff)
+        return other.tensor(self)
+
+    def compose(self, other, qargs=None, front=False):
+        """Return the composed operator.
+
+        Args:
+            other (BaseOperator): an operator object.
+            qargs (list or None): a list of subsystem positions to apply
+                                  other on. If None apply on all
+                                  subsystems [default: None].
+            front (bool): If True compose using right operator multiplication,
+                          instead of left multiplication [default: False].
+
+        Returns:
+            BaseOperator: The operator self @ other.
+
+        Raises:
+            QiskitError: if other has incompatible dimensions for specified
+                         subsystems.
+
+        Additional Information:
+            Composition (``@``) is defined as `left` matrix multiplication for
+            matrix operators. That is that ``A @ B`` is equal to ``B * A``.
+            Setting ``front=True`` returns `right` matrix multiplication
+            ``A * B`` and is equivalent to the :meth:`dot` method.
+        """
+        if qargs is None:
+            qargs = getattr(other, 'qargs', None)
+
+        if not isinstance(other, BaseOperator):
+            other = Operator(other)
+
+        input_dims, output_dims = self._get_compose_dims(other, qargs, front)
+
+        # If other is also an ScalarOp we only need to possibly
+        # update the coefficient and dimensions
+        if isinstance(other, ScalarOp):
+            if self.coeff is None:
+                coeff = other.coeff
+            elif other.coeff is None:
+                coeff = self.coeff
+            else:
+                coeff = self.coeff * other.coeff
+            return ScalarOp(input_dims, coeff=coeff)
+
+        # If we are composing on the full system we return the
+        # other operator with reshaped dimensions
+        if qargs is None:
+            ret = other.reshape(input_dims, output_dims)
+            if self.coeff is None or self.coeff == 1:
+                return ret
+            return self.coeff * ret
+        # Otherwise compose using other operators method
+        # Note that in this case that operator must know how to initalize
+        # from an ScalarOp either using its to_operator method or having
+        # a specific initialization case.
+        return other.__class__(self).compose(
+            other, qargs=qargs, front=not front)
```

Comment:
can you explain what's going on here? if no `qargs` are specified during compose, you compose on the full subsystem.. so qargs would become equal to `other.qargs` before reaching this point. would that work if other is smaller or larger than self?

Reply:
i see erick asked a similar question, but i just couldn't follow from the code.

Reply:
This allows the constructor for another BaseOperator subclass to handle conversion of the identity (or scalar identity) to whatever data format it has, and then composes the other operator with it.

Eg if it was a matrix it would convert to the identity matrix and then compose.
If it was a Choi it would convert to identity channel Choi matrix then compose
If it was a Clifford it would convert to identityt Clifford then compose etc...

Reply:
I think the `front= not front` is a typo, not that order makes a difference for scalar op.